controllers/networking-calico/pkg/charts/utils.go

```diff
@@ -27,68 +30,168 @@ const (
-				calico.TyphaImageName:                                      imagevector.CalicoTyphaImage(),
-				calico.KubeControllersImageName:                            imagevector.CalicoKubeControllersImage(),
-				calico.NodeImageName:                                       imagevector.CalicoNodeImage(),
-				calico.PodToDaemonFlexVolumeDriverImageName:                imagevector.CalicoFlexVolumeDriverImage(),
-				calico.TyphaClusterProportionalAutoscalerImageName:         imagevector.TyphaClusterProportionalAutoscalerImage(),
-				calico.TyphaClusterProportionalVerticalAutoscalerImageName: imagevector.TyphaClusterProportionalVerticalAutoscalerImage(),
-			},
-			"global": map[string]string{
-				"podCIDR": network.Spec.PodCIDR,
-			},
-			"ipip": calicov1alpha1.Always,
-		}
-		calicoConfigValues = map[string]interface{}{
-			"backend": calicov1alpha1.Bird,
-		}
-		ipamConfig = map[string]interface{}{
-			"type":   hostLocal,
-			"subnet": usePodCIDR,
-		}
-		typhaConfig = map[string]interface{}{
-			"enabled": true,
-		}
-	)
+func ComputeCalicoChartValues(network *extensionsv1alpha1.Network, config *calicov1alpha1.NetworkConfig) (map[string]interface{}, error) {
+	typedConfig, err := generateChartValues(config)
+	if err != nil {
+		return nil, fmt.Errorf("error when generating calico config: %v", err)
+	}
+	calicoConfig, err := typedConfig.toMap()
+	if err != nil {
+		return nil, fmt.Errorf("could not convert calico config: %v", err)
+	}
+	calicoChartValues := map[string]interface{}{
+		"images": map[string]interface{}{
+			calico.CNIImageName:                                        imagevector.CalicoCNIImage(),
+			calico.TyphaImageName:                                      imagevector.CalicoTyphaImage(),
+			calico.KubeControllersImageName:                            imagevector.CalicoKubeControllersImage(),
+			calico.NodeImageName:                                       imagevector.CalicoNodeImage(),
+			calico.PodToDaemonFlexVolumeDriverImageName:                imagevector.CalicoFlexVolumeDriverImage(),
+			calico.TyphaClusterProportionalAutoscalerImageName:         imagevector.TyphaClusterProportionalAutoscalerImage(),
+			calico.TyphaClusterProportionalVerticalAutoscalerImageName: imagevector.TyphaClusterProportionalVerticalAutoscalerImage(),
+		},
+		"global": map[string]string{
+			"podCIDR": network.Spec.PodCIDR,
+		},
+		"config": calicoConfig,
+	}
+	return calicoChartValues, nil
+}
 
-	if config != nil {
-		switch config.Backend {
-		case calicov1alpha1.Bird, calicov1alpha1.None:
-			calicoConfigValues["backend"] = config.Backend
+func generateChartValues(config *calicov1alpha1.NetworkConfig) (*calicoConfig, error) {
+	c := newCalicoConfig()
+	if config == nil {
+		return &c, nil
+	}
+
+	switch config.Backend {
+	case calicov1alpha1.Bird, calicov1alpha1.VXLan, calicov1alpha1.None:
+		c.Backend = config.Backend
+	default:
+		return nil, fmt.Errorf("unsupported value for backend: %s", config.Backend)
```

Comment:
I have legacy shoot cluster that has the following calico configuration(undefined `backend`).
```yaml
networking:
  nodes: 10.250.0.0/19
  pods: 100.96.0.0/11
  providerConfig:
    apiVersion: calico.networking.extensions.gardener.cloud/v1alpha1
    kind: NetworkConfig
  services: 100.64.0.0/13
  type: calico
```

Now I tried to upgrade to this version of the extension and I got the following error 
```bash

msg="Network did not get ready yet" error="extension encountered error during reconciliation: Error reconciling network: error when generating calico config: unsupported value for backend: " operation=reconcile shoot=garden-dev/local-shoot
```

Please, adapt your changes so that legacy clusters to be still compatible?

Reply:
I agree, this configuration should not fail. I fixed this by declaring the backend optional as well. Would you please try this again for me?

What I want to point out is that this implementation will fail when a user provides invalid configuration inputs, which was silently ignored before. This could potentially also break backwards-compatibility, but I think it's worth to break there because this indicates a user providing a configuration that was never really reflected by Calico.

Reply:
Yes, agreed, thanks Gerrit.

Reply:
Yes, agreed, thanks Gerrit.